### PR TITLE
[webview_flutter] Adds platform implementations for onHttpError callback

### PR DESCRIPTION
**Pull request by HugoOlthof**
Reference https://github.com/flutter/packages/pull/3278

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by stuartmorgan** [comment]
We should make sure https://github.com/flutter/packages/pull/3675 gets the baseline file updates and then lands, and then this has that merged in, before this lands, so that we can better see what the baseline changes here are and if any new warnings are being introduced.

This doesn't have to block initial review, but I'm flagging it now so that we don't forget.

**Comment by bparrishMines** [comment]
https://github.com/flutter/packages/pull/3675 has been merged, so this PR will need to pull from main and handle the Android lint warnings.

File: packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFDataConverters.h
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFDataConverters.h:113]: nit: We recently decided on a new naming scheme for these to include the word `Native` to make it easier to read. Can you change this to `FWFWKNavigationResponseDataFromNativeNavigationResponse`

```suggestion
extern FWFWKNavigationResponseData *FWFWKNavigationResponseDataFromNativeNavigationResponse(
```

File: packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFDataConverters.h
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFDataConverters.h:122]:
```suggestion
extern FWFNSHttpUrlResponseData *FWFNSHttpUrlResponseDataFromNativeNSURLResponse(NSURLResponse *response);
```

File: packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFDataConverters.h
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFDataConverters.h:150]: And this to
```suggestion
extern WKNavigationResponsePolicy FWFNativeWKNavigationResponsePolicyFromEnumData(
```

File: packages/webview_flutter/webview_flutter_wkwebview/CHANGELOG.md
Comment (bparrishMines) [inline packages/webview_flutter/webview_flutter_wkwebview/CHANGELOG.md:35]: I think this was added by accident from pulling main?
```suggestion
```

**Review by bparrishMines — APPROVED** [review]
LGTM with last few nits

File: packages/webview_flutter/webview_flutter_android/example/integration_test/webview_flutter_test.dart
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_android/example/integration_test/webview_flutter_test.dart:915]: This needs to be `primaryUrl`, not an actual site. Using real sites in integration tests makes tests much more prone to flake.

File: packages/webview_flutter/webview_flutter_wkwebview/example/integration_test/webview_flutter_test.dart
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_wkwebview/example/integration_test/webview_flutter_test.dart:977]: `primaryUrl`

File: packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFNavigationDelegateHostApi.m
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFNavigationDelegateHostApi.m:184]: This has the some problem as the code discussed recently in the permission callback PR; it will unconditionally call the callback with an invalid value if Apple ever adds a new enum value. @bparrishMines Are you going to mass-fix this in a fast follow? I'm concerned about replicating a pattern that we know is dangerous.

File: packages/webview_flutter/webview_flutter_android/lib/src/android_webview_controller.dart
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_android/lib/src/android_webview_controller.dart:760]: I'm confused; how is this callback intended to work in practice? You are throwing away the request object before it gets to the client, which means they will have absolutely no way of knowing whether the error is coming from a navigation (and if so to where) or a resource on the page (and if so, what). I'm not clear on what non-trivial use case a client could satisfy without having the request.

Am I missing something here?

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/webview_flutter/webview_flutter_android/lib/src/android_webview_controller.dart
Comment (stuartmorgan) [reply packages/webview_flutter/webview_flutter_android/lib/src/android_webview_controller.dart:760]: /cc @bparrishMines who reviewed the overall PR, since it's core to the overall design. (It looks like the main PR only had one review before sub-parts started landing? Ideally that should have been two.)

Comment (HugoOlthof) [reply packages/webview_flutter/webview_flutter_android/lib/src/android_webview_controller.dart:760]: It makes sense to include the request in the HttpResponseError object. Is it better to create a platform specific implementation for HttpResponseError and include the request in there, like AndroidWebResourceError, or include it in the HttpResponseError object itself?

Comment (stuartmorgan) [reply packages/webview_flutter/webview_flutter_android/lib/src/android_webview_controller.dart:760]: Since it seems core to using this correctly, I would include it in the base object; it could be nullable in case there are implementations or cases where it's not available.

**Comment by HugoOlthof** [comment]
@stuartmorgan @bparrishMines I've added a new PR https://github.com/flutter/packages/pull/4025 which adds the request to the HttpResponseError object.

File: packages/webview_flutter/webview_flutter_android/lib/src/android_webview_controller.dart
Comment (bparrishMines) [reply packages/webview_flutter/webview_flutter_android/lib/src/android_webview_controller.dart:760]: @stuartmorgan Sorry about that. I've been reviewing all of the WebView PRs as the owner of both the Android and iOS implementation, so I assumed it was fine to be the only approval for the aggregate PR. I'll start getting secondary reviews for those as well.

I agree with creating adding a request object to `HttpResponseError`.

File: packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFNavigationDelegateHostApi.m
Comment (bparrishMines) [reply packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFNavigationDelegateHostApi.m:184]: This is fixed now with https://github.com/flutter/packages/pull/3812. 

This also only affects enums that we receive from Objective-C. The Dart code shouldn't be able to return an enum the native side doesn't recognize.

File: packages/webview_flutter/webview_flutter_android/lib/src/android_webview_controller.dart
Comment (bparrishMines) [reply packages/webview_flutter/webview_flutter_android/lib/src/android_webview_controller.dart:760]: It actually looks like Android and iOS have a request object and a response object. 

Android:
https://developer.android.com/reference/android/webkit/WebResourceRequest
https://developer.android.com/reference/android/webkit/WebResourceResponse

iOS:
https://developer.apple.com/documentation/foundation/nsurlrequest?language=objc
https://developer.apple.com/documentation/foundation/nshttpurlresponse

In their respective methods,
https://developer.apple.com/documentation/webkit/wknavigationdelegate/1455643-webview?language=objc
https://developer.android.com/reference/android/webkit/WebViewClient#onReceivedHttpError(android.webkit.WebView,%20android.webkit.WebResourceRequest,%20android.webkit.WebResourceResponse)

Android provides the request and response and iOS provides the response. I think we could add a nullable `LoadRequestParams` to `HttpResponseError` AND a new nullable `WebResourceResponse` field.

**Comment by stuartmorgan** [comment]
Update from triage: the response object PR, which is blocking this, is still in progress.

**Comment by stuartmorgan** [comment]
Update from triage: still waiting for https://github.com/flutter/packages/pull/4025, then this can proceed.

**Comment by Hixie** [comment]
This is blocked on https://github.com/flutter/packages/pull/4025 which is currently in review.

**Comment by stuartmorgan** [comment]
Update from triage: #5790 has been created as a copy of the blocking PR, to work on unblocking this.

**Comment by stuartmorgan** [comment]
Update from triage: the blocker has been landed, so this is ready to be updated.

**Comment by stuartmorgan** [comment]
Closing in favor of the PR cross-linked just above, as this one isn't editable by the Flutter team.

**Pull request closed without merging**
